edx_event_bus_kafka/internal/consumer.py

```diff
@@ -282,11 +287,88 @@ def record_event_consuming_error(self, run_context, error, maybe_event):
+            # .. custom_attribute_description: Boolean describing whether there was a message involved. Some errors
+            #   have no message.
+            set_custom_attribute('kafka_has_message', bool(kafka_message))
+
+            if kafka_message:
+                headers = kafka_message.headers() or []  # treat None as []
+                # header is list of tuples, so handle case with duplicate headers for same key
+                message_ids = [value.decode("utf-8") for key, value in headers if key == EVENT_ID_HEADER]
+                if message_ids and len(message_ids) > 0:
+                    # .. custom_attribute_name: kafka_message_id
+                    # .. custom_attribute_description: The message id which can be matched to the logs. Note that the
+                    #   header in the logs will use 'ce_id'.
+                    set_custom_attribute('kafka_message_id', ",".join(message_ids))
+
+            if kafka_error:
+                # .. custom_attribute_name: kafka_error_fatal
+                # .. custom_attribute_description: Boolean describing if the error is fatal.
+                set_custom_attribute('kafka_error_fatal', kafka_error.fatal())
+                # .. custom_attribute_name: kafka_error_retriable
+                # .. custom_attribute_description: Boolean describing if the error is retriable.
+                set_custom_attribute('kafka_error_retriable', kafka_error.retriable())
+
+        except Exception as e:  # pylint: disable=broad-except, pragma: no cover
+            # Use this to fix any bugs in what should be benign monitoring code
+            set_custom_attribute('kafka_monitoring_error', repr(e))
+
+    def _get_kafka_message_and_error(self, message, error):
+        """
+        Returns tuple of (kafka_message, kafka_error), if they can be found.
+
+        Notes:
+            * If the message was sent as a parameter, it will be returned.
+            * If the message was not sent, and a KafkaException was sent, the
+                message will be pulled from the exception if it exists.
+            * A KafkaError will be returned if it is either passed directly,
+                or if it was wrapped by a KafkaException.
+
+        Arguments:
+            message: None if event could not be fetched or decoded, or a Message if one
+              was successfully deserialized but could not be processed for some reason
+            error: An exception instance, or None if no error.
+        """
+        if not error or isinstance(error, KafkaError):
+            return message, error
+
+        kafka_error = getattr(error, 'kafka_error', None)
+        if not kafka_error and len(error.args) > 0 and isinstance(error.args[0], KafkaError):
+            # KafkaException uses args[0] to wrap the KafkaError
+            kafka_error = error.args[0]
+
+        if message:
+            # give priority to the passed message, although it should be the same message, in theory
```

Comment:
Should we record somewhere if the messages didn't match?

Reply:
Done.